### DISCUSSION TITLE
[DataImporter] Two fidi docker container instances for two different users cross-loading data

### DISCUSSION THREAD

**Discussion by diamant-x** [firefly-iii/firefly-iii#5998] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

I have one Firefly3 setup with two different users, i want each of them to have an automatic fidi to import their independent transactions.

I have setup already Fidi for one of the users, with its own Personal Access Token.
I have setup Fidi with a unique set of Nordigen app keys.
Nordigen import works fine and when creating the config, the dropdown of firefly accounts matches those available for user1.

I have deployed a new FiDi container, with a separate config folder and instance.
I have setup FiDi with a new Personal Access Token under user2. The tokens are clearly different between those of user1 and user2.
I have setup FiDi2 with new Nordigen app keys from the same organization, so different app keys from user1.
When creating a new nordigen config flow in the new container for user2, the lsit of available accounts are the same as those for user1, and those of user2 are not available at all.
user2 uses the same bank and bankaccount 
user2 is a FF3 admin account.

Thanks!

### Expected behaviour

FiDi from user1 sees its own firefly accounts only
FiDi from user2 sees its own firefly accounts only.

### Debug information

Can't find any specific mention about the connection to firefly3 from fidi, all the logs are related to the (succesful) connection to nordigen.

### Steps to reproduce

1. setup FF3 with 2 users
2. seutp different asset accounts for each user
3. setup one fidi for user1 with personal acces token
4. confirm fidi for user1 shows accounts for user1 only
5. seutp a second fidi for user 2 with its personal access token
6. check the available account to be mapped
error occurs, accounts from user 1 appear.

**Comment by JC5** [comment]
The (encrypted) access token is cached in a cookie, and could also be present in the `FIREFLY_III_ACCESS_TOKEN` environment variable. If the two data importers share the same (sub)domain, the information from one FIDI instance can leak to the other one.

**Comment by JC5** [comment]
> Can't find any specific mention about the connection to firefly3 from fidi, all the logs are related to the (succesful) connection to nordigen.

Look for stuff like this:

```
[2022-04-13 14:42:17] local.DEBUG: Now at App\Http\Controllers\TokenController::index  
[2022-04-13 14:42:17] local.DEBUG: No access token in hasAccessToken(), will return config variable.  
[2022-04-13 14:42:17] local.DEBUG: No client id in hasClientId(), will return config variable.  
[2022-04-13 14:42:17] local.DEBUG: No base url in getBaseUrl(), will return config variable.  
[2022-04-13 14:42:17] local.DEBUG: No vanity url in getVanityUrl(), will return config variable.  
[2022-04-13 14:42:17] local.INFO: The following configuration information was found:  
[2022-04-13 14:42:17] local.INFO: Personal Access Token: "" (limited to 25 chars if present)  
[2022-04-13 14:42:17] local.INFO: Client ID            : "0"  
[2022-04-13 14:42:17] local.INFO: Base URL             : "http://firefly.jc.home"  
[2022-04-13 14:42:17] local.INFO: Vanity URL           : "http://firefly.jc.home"  
[2022-04-13 14:42:33] local.DEBUG: Now at App\Http\Controllers\TokenController::submitClientId  
[2022-04-13 14:42:33] local.DEBUG: Submitted data:  {"client_id":"4","base_url":"https://firefly.jc.home/"} 
[2022-04-13 14:42:33] local.DEBUG: Base URL is "http://firefly.jc.home"  
[2022-04-13 14:42:33] local.DEBUG: Vanity URL is now "http://firefly.jc.home"  
[2022-04-13 14:42:33] local.DEBUG: Base URL is now "https://firefly.jc.home/"  
[2022-04-13 14:42:33] local.DEBUG: Now in App\Http\Controllers\TokenController::redirectForPermission(request, "https://firefly.jc.home", "http://firefly.jc.home", 4)  
[2022-04-13 14:42:33] local.DEBUG: Query parameters are {"client_id":4,"redirect_uri":"https://firefly-data.jc.home/callback","response_type":"code","scope":"","state":"UPxnrSwGEp4lAOspwLpRcw1E2oJY7FyAP2QNITky","code_challenge":"d9kcEJhV6k_vwSOYorDmfrvlH5kexzmaB_-nuVtOAR4","code_challenge_method":"S256"} 
[2022-04-13 14:42:33] local.DEBUG: Now redirecting to "http://firefly.jc.home/oauth/authorize?" (params omitted)  
[2022-04-13 14:42:34] local.DEBUG: Now at App\Http\Controllers\TokenController::callback  
[2022-04-13 14:42:34] local.DEBUG: State is valid!  
[2022-04-13 14:42:34] local.DEBUG: Params for access token {"form_params":{"grant_type":"authorization_code","client_id":4,"redirect_uri":"https://firefly-data.jc.home/callback","code_verifier":"1m...e"}} 
[2022-04-13 14:42:34] local.DEBUG: Will contact "https://firefly.jc.home/oauth/token" for a token.  
[2022-04-13 14:42:36] local.DEBUG: Response {"token_type":"Bearer","expires_in":1209598,"access_token":"eyJ0eXAi...b902169c65e79078c0869"} 
[2022-04-13 14:42:36] local.DEBUG: Return redirect with cookies to "https://firefly-data.jc.home"  


```

**Comment by diamant-x** [comment]
> The (encrypted) access token is cached in a cookie, and could also be present in the FIREFLY_III_ACCESS_TOKEN environment variable. If the two data importers share the same (sub)domain, the information from one FIDI instance can leak to the other one.

Both FiDis are deployed in the same docker host, so they share the same subdomain, one is working on port 3474 and the other in port 3475. I'm not sure to understand how that would be an issue?

This is the log i'm getting. I do see that the beginning of the personal access token is identical for both users, not sure if that might clash?
```
[2022-04-13 14:58:01] production.DEBUG: Now at App\Http\Controllers\IndexController::reset  
[2022-04-13 14:58:01] production.INFO: The following configuration information was found:  
[2022-04-13 14:58:01] production.INFO: Personal Access Token: "<redacted-token>" (limited to 25 chars if present)  
[2022-04-13 14:58:01] production.INFO: Client ID            : "0"  
[2022-04-13 14:58:01] production.INFO: Base URL             : "http://mydomain:3473"  
[2022-04-13 14:58:01] production.INFO: Vanity URL           : "http://mydomain:3473"  
[2022-04-13 14:58:01] production.DEBUG: Found personal access token + URL "http://mydomain:3473" in config, set cookie and return to index.  
192.168.10.26 - - [13/Apr/2022:14:58:01 +0200] "GET /token HTTP/1.1" 302 4651 "http://mydomain:3475/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.75 Safari/537.36 Edg/100.0.1185.39"
[2022-04-13 14:58:01] production.DEBUG: Now in App\Http\Controllers\IndexController::index  
[2022-04-13 14:58:01] production.DEBUG: App\Services\Shared\Authentication\SecretManager::hasValidSecrets  
```

I see that my logs don't include the 'Response' from the FF3 like yours. any way to force that to appear?

**Reply by JC5** [reply]
Make sure that FIDI 2 doesn't have the `FIREFLY_III_ACCESS_TOKEN` and try accessing it from a incognito browser

**Reply by diamant-x** [reply]
I did these two things and it seems to worked fine now, thanks!

- Would you say is better to use the OAuth method without using the ACCESS_TOKEN method? In scenarios with multiple FiDis running?
- Does the incognito mode is needed everytime a manual import is done through the FiDi webUI?

Thanks.

**Reply by JC5** [reply]
OAuth is better yes, if you set up only the `FIREFLY_III_CLIENT_ID` and the `FIREFLY_III_URL`.

It may be smart to make sure that both importers are on separate subdomains so the cookies don't clash.

**Reply by diamant-x** [reply]
> It may be smart to make sure that both importers are on separate subdomains so the cookies don't clash.

Is this a hard requirement or is there any way to make it work in the same subdomain.?
I have them both on the same docker cluster so they only distinguish between themselves based on the port, as the server is the same for both. i tried configuring one using the IP and the other the hostname to try to avoid the domain issues but doesn't seem to like it as now the one i used OAuth to connect is displaying the error 
```
Could not connect to Firefly III: Error accessing "http://192.168.1.10:3473/api/v1/about?". Status code is 401. Body is: {"message":"Unauthenticated","exception":"AuthenticationException"}
Could not connect to your local Firefly III instance.
```
I will try to configure it again using the internal docker IPs, let's see if it stays that way.

**Reply by diamant-x** [reply]
> I will try to configure it again using the internal docker IPs, let's see if it stays that way.

Apparently the pre-authorization is also storing the callback url so i can't change the OauthID callback url and expect that the already authorized items will go to the new url, so i can't replace it to use the docker-url.

So my current status is FiDi1 says that my nordigen accounts ids are not found anymore, 
and FiDi 2 says my OAuth is no longer authorized.

I did the configuration for both one in normal browser and the other in incognito, one with hostname and the other with LAN IP.
Seems like:
1) the second FiDi connection to the same nordigen bank username 'disabled' the FiDi1 nordigen unique accounts Ids. Is this a hard constrain? One Nordigen connection can't be used in two FiDis?
2) although one FiDi uses hostname and the other IP, the authorization breaks up

**Reply by JC5** [reply]
> Is this a hard requirement or is there any way to make it work in the same subdomain.?

I'm afraid not, but I'm open to make the cookie name configurable. Would that be something you'd like to try out? 

> the second FiDi connection to the same nordigen bank username 'disabled' the FiDi1 nordigen unique accounts Ids. Is this a hard constrain? One Nordigen connection can't be used in two FiDis?

Probably a security measure from Nordigen's side, so if you steal somebody's account identifiers you can't easily abuse them.

**Reply by diamant-x** [reply]
> > Is this a hard requirement or is there any way to make it work in the same subdomain.?
> 
> I'm afraid not, but I'm open to make the cookie name configurable. Would that be something you'd like to try out?

I'm fine testing it out. I'm really interested on finding a way to make this work with 2 (or more) separate accounts. Also open to other approaches if any. Many thanks!

I'm running FiDi through a container configured by this other (active) developer so changes might take a while to communicate. https://github.com/alexbelgium/hassio-addons

**Reply by JC5** [reply]
Sorry, customizable cookie names are out of reach for now. If you have a reverse proxy, you could set up different subdomains perhaps?

**Reply by diamant-x** [reply]
I just setup nginx proxy Manager, and remapped to two different subdomains the corresponding FiDis. Firefly3 still sits in the same common domain as both accounts are part of the same Firefly platform.
So far i managed to restablish the authentication (Despite the proxy maanger redirecting the port 80 to the 34xx of each one, the callback failed if I didn't use the explicit port exposed by the container).

I'll restart the authenticated FiDi containers and leave it there for the cron job to pick up the import and see if it works fine now or not.

**Reply by diamant-x** [reply]
Ok, no luck with setting up subdomain proxys either...

Both FiDi1 and FiDi2 errored with the same error:

```
"Could not connect to Firefly III: Error accessing "http://firefly3.domain.ml/api/v1/about?". Status code is 401. Body is: {"message":"Unauthenticated","exception":"AuthenticationException"}
Could not connect to your local Firefly III instance."
```
Apparently, when I manually navigate through the proxied subdomain in the web browser, it all works fine.
But when running through cron, the connection is stablished from the original domain:port, so the authentication fails because it is not initiated from the subdomain it was authenticated first in the browser and the callback fails.

Did you successfully achieve 2 FiDis working with some specific architecture you could share?
Thanks.

**Reply by diamant-x** [reply]
> Did you successfully achieve 2 FiDis working with some specific architecture you could share?

Do you have any valid working setup of 2 FiDis? Or what is your recommended approach to multiple accounts importers?

Thanks.

**Reply by diamant-x** [reply]
> > Did you successfully achieve 2 FiDis working with some specific architecture you could share?
> 
> Do you have any valid working setup of 2 FiDis? Or what is your recommended approach to multiple accounts importers?
> 
> Thanks.

Still interested on knowing this.

As of now, I'm unable to have FiDi auto-import anymore. Seems like having different subdomains messes it up with this system. Manually doing it still works though.

**Reply by JC5** [reply]
It's still on my list, but I haven't found the time to set it all up.

**Reply by diamant-x** [reply]
Thanks for the followup. I have since been unable to get the auto-import to work anymore with just one FiDi anyways. Manually doing the steps works fine. When being triggered on its own through cron (within docker), it fails under Authentication failure. Tried with the vanity set to my subdomain and different IPs attemps, but they all fail the same way. I think when running through cron in docker, Communications go through the internal docker network and then the callback url in FF3 fails to match the previously authorized one.

**Reply by JC5** [reply]
I was able to get a single instance to work as if it were two, by visiting one in a private browser window. For this experiment, I used a reverse proxy to give them different names but I'm convinced I can do the same with the same host name.

<img width="1197" alt="Screenshot 2023-04-10 at 14 14 35" src="https://user-images.githubusercontent.com/5889984/230900065-5ee3e44a-b690-489c-9194-aa44de8a65e5.png">

**Reply by diamant-x** [reply]
Thanks for the test. So just to understand (correct me if i'm wrong).

1 FF3 instance.
1 FiDi instance.
2 reverse proxy hostname (different one to another) pointing to the fidi instance and port.

in the private browser, you had to a) login to FF3 instance with second account and b) authorize the FiDi instance through the second proxy hostname as callback url?

What authorization method did you use?

Thanks!

**Reply by JC5** [reply, marked as answer]
> in the private browser, you had to a) login to FF3 instance with second account and b) authorize the FiDi instance through the second proxy hostname as callback url?

Yes. Both needed authentication separately. I set no authentication variables at all. Each user had to input their client ID and the Firefly III url.

**Reply by diamant-x** [reply]
I'm just coming back to this, as I have now a proper proxy setup I can use to try subdomains and such.
However, I'm realizing that by removing the authentication variables, then FiDi wouldn't be able to auto-import to a 'default' account.

- I guess ideally the auto_import config files could/should store authentication variables for each user so that a single FiDi instance could work and autoimport different configs for different user profiles.

Or at least a way for FiDi to just, 'ignore' any authentication variables so that it can run as an 'incognito mode' and be able to use it for sporadical loads to a secondary account.

Maybe I misunderstood something but something like that would be great to properly enable Firefly3 for multi-users.

EDIT: Or maybe to be able for the client ID to be defined by the http headers on the reverse proxy, so that I could set the subdomains to add the client ID which they should use and use those.... Although I guess that wouldn't help for auto-import as it would run without any http query.

EDIT2: Just to add to the community, I was able to do it just as mentioned in the last answer. No authentication variables. 2 different subdomains pointing to the same container. Opening one FF3 session in incognito (or different browser) and one of the subdomains as well. Setting the client id as not-confidential and with the chosen subdomain as callback.

**Comment by diamant-x** [comment]
Another approach I [saw in the documentation](https://docs.firefly-iii.org/data-importer/install/configure/#client-id-firefly-iii-url) was 
`You can generate your own client ID on your Profile page (under OAuth). This is the ID you need when you want to share FIDI with multiple people, or when you want to allow others to use the same instance of FIDI.`

- How would the architecture work for this scenario to happen? 
- Is there a way to automate the cron jobs for different users sharing the same FiDi? Or does this documentation description only apply to manual web-browser-based imports?

Thanks.

**Reply by JC5** [reply]
If it's on the command line, you can take the two tokens and run the jobs.
